Dask image integration

michalk8 opened the pull request:
### Types of changes
<!--- What types of changes are introduced? Put an `x` in all applicable boxes: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (change that would cause existing functionality to not work as before)
- [x] New feature (non-breaking change which adds functionality)

### Description
<!--- What does this PR achieve? -->
Add dask support.

### How has this been tested?
<!--- Describe how you've tested your changes: -->
Many times.

### Closes
<!--- If applicable, add `closes #XXXX` below to auto-close the issue related to this PR: -->
closes #296 
closes #327 
closes #328 
closes #333 
closes #336 
closes #337

michalk8 commented:
@giovp First version of watershed using dask is done, see below the comparison (chunks/size=1000):
current (note the vertical line in the middle for both labels and the image - chunking artifacts for preproc./segmentation)
![2021-04-29T19:24:39+02:00](https://user-images.githubusercontent.com/46717574/116595167-54684080-a923-11eb-8136-911c9d2bfde6.png)
map_blocks (note that there's a vertical line in the middle - chunking artifact for segmentation)
![2021-04-29T19:07:12+02:00](https://user-images.githubusercontent.com/46717574/116595173-55996d80-a923-11eb-821c-9b2fc405fd91.png)
map_overlap (note no such vertical line)
![2021-04-29T19:03:29+02:00](https://user-images.githubusercontent.com/46717574/116595176-56320400-a923-11eb-90a7-8770daae9c1c.png)

Performance-wise, it took 10s, compared to 20s using 8 cores on my machine, will need to benchmark this more thouroughly though, but the runtime seems fine.
However, there are other design decisions to make:

- [ ] remove segmentation blob? (don't really want to port it to dask
- [ ] how to handle custom models? before, the crops were passed (possibly in parallel) to the function
  we can keep the old cropping approach, but it has drawbacks:
  1. boundaries - we could do `dask.array.map_overlap`, as it's done for `watershed` right now
  2. DL-models + thread/process parallelization won't play together nicely - if above is used, we could rely on user not to parallelize over chunks, TF/PT models should be fine w.r.t. mem/cores
  relate to #326 
- [ ]  currently, `ImageContainer.apply` handles lazy computation very clumsily, always calling compute after getting the result
  we could potentially lazy-chain some computations, possible ideas:
  1. LazyImageContainer - does everything lazily unless you call `.compute()` or e.g. napari needs the data
  2. expose `lazy: bool = ...` to `ImageContainer.apply` + high level methods like `squidpy.im.process`
  3. create `apply_lazy` which computes stuff lazily, whereas if a lazy array is supplied to `.apply`, it is computed
  currently, I'd go for 1 or 2 (think 1st is a bit too heavy approach)

giovp commented:
> remove segmentation blob? (don't really want to port it to dask

yes sounds great

>  how to handle custom models? before, the crops were passed (possibly in parallel) to the function
>  boundaries - we could do dask.array.map_overlap, as it's done for watershed right now
DL-models + thread/process parallelization won't play together nicely - if above is used, we could rely on user not to parallelize over chunks, TF/PT models should be fine w.r.t. mem/cores
relate to #326

I agree that we should rely on user to not parallelize, and yse agree with map overlap...

>  currently, ImageContainer.apply handles lazy computation very clumsily, always calling compute after getting the result
we could potentially lazy-chain some computations, possible ideas:
LazyImageContainer - does everything lazily unless you call .compute() or e.g. napari needs the data
expose lazy: bool = ... to ImageContainer.apply + high level methods like squidpy.im.process
create apply_lazy which computes stuff lazily, whereas if a lazy array is supplied to .apply, it is computed
currently, I'd go for 1 or 2 (think 1st is a bit too heavy approach)

I would go for 2 as well, this essentially mean exposing lazy to all `process` and `segment` right? and for napari not necessary cause is able to handle lazyness right?
@hspitzer thoughts on this?

michalk8 commented:
@giovp @hspitzer I think I've finished the integration, would you mind testing it/giving feedback as well?

To summarize the changes:
- using dask/dask_image whereever possible to avoid artifacts when chunking
- add `channelwise: bool = True` to `ImageContainer.show`
- add `ImageContainer.__setitem__`, .e.g `img['foo'] = some_array`, just for convenience at the cost of some control
- add `labels` layer to napari (use 
- allow channels to be shown separately in napari (automatically detected, can be set as `img[<layer>].attrs['luminance'] = True/False`
- allow `channel_dim=None` in `add_img` (creates automatically iff name/dim mismatch happens, e.g. `channels_0`)
- add lazy loading on png/jpegs
- add `ImageContainer.rename` to rename a layer (could also extend this for channels)
- add tab completion for ImageContainer (try tab-completing `img['im`)
- remove `SegmentationBlob`

TODOs:
- [ ] deal with `pims` + very large images (thinks it's a zip bomb)
- [ ] test the new additions/changes + feedback
- [ ] deal with non Z-dim TODOs in the code
- [ ] update docstrings
- [ ] write tests

Here's some starting code I've used, but more thorough testing is needed (lazy=True, seems to work as well):
```python
import squidpy as sq
import numpy as np
import napari

img = sq.datasets.visium_fluo_image_crop()
adata = sq.datasets.visium_fluo_adata_crop()
sq.im.process(img, chunks=1000, method='smooth', sigma=[1, 1, 0])
img.show("image_smooth", channelwise=True, dpi=180, cmap='gray')

sq.im.segment(img, layer='image_smooth', chunks=1000)

with napari.gui_qt():
    img.interactive(adata)
```

hspitzer commented on squidpy/im/_container.py:
```diff
@@ -172,14 +177,16 @@ def add_img(
         Parameters
         ----------
         img
-            In memory array or path to on-disk *TIFF*/*JPEG* image.
```
the general ImageContainer doctoring needs to be updated as well

hspitzer commented on squidpy/im/_process.py:
```diff
@@ -74,22 +90,26 @@ def process(
     if channel_dim is None:
         channel_dim = img[layer].dims[-1]
     layer_new = Key.img.process(method, layer, layer_added=layer_added)
+    map_kwargs: Dict[str, Any] = {"lazy": lazy}
```
allow passing of `depth` to `map_kwargs`?

hspitzer reviewed (commented):
Great work @michalk8 
Its lovely to see all the dask image functionalities here — especially the possibility of using `map_overlap` to take care of crop boundaries. And the tab-completion is amazing!! How is this implemented?

Ok, for removing SegmentationModelBlob. It was not very good anyways. I’d rather have support for Stardist etc in the future than some old models that no one uses.

I’ve played around with the functions a bit and have a couple of comments:
* in `im.process`, there is no way to pass kwargs to `map_overlap`, because all kwargs are passed to the processing function. I could imagine a use-case though, when you have a custom processing function, and would like to use the overlap functionalities. Could you expose the `map_overlap` arguments somehow (or at least a couple of them)?
* for the `SMOOTH` processing, I did not quite understand how this is implemented. You are setting chunks to 0, before passing the gaussian filter to `img.apply`. Why? Could you explain what is happening there?
* the new segmentation is great as well. I observe less border artifacts! I noticed, that you can see the chunks structure from the labels, and was wondering if you have tested somehow that cells are not labelled with the same label? 
* is it possible to tell dask to use several processes when applying functions to chunks?
* did you test the lazy loading for big pngs/jpegs? I don’t have a large image like this atm to test this.

And just as a reminder (also for myself): we should update with docs with these changes. In particular changes are needed in 
* ImageContainer example
* processing, segmentation example with pointers to how borders between chunks are handled
* we could add an example for `img.show` which now has some nice functionalities

michalk8 commented:
> Could you expose the map_overlap arguments somehow (or at least a couple of them)?

Good idea, will do that.

> You are setting chunks to 0, before passing the gaussian filter to img.apply. Why? Could you explain what is happening there?

That's because `dask-image`'s `gaussian_filter` already contains `map_overlap` inside. In `ImageContainer.apply`, if `chunks != None` controls whether we use dask or not.

> I noticed, that you can see the chunks structure from the labels, and was wondering if you have tested somehow that cells are not labelled with the same label?

Yes, there's this [test](https://github.com/theislab/squidpy/pull/324/files#diff-7c41484d346aac7a9605c1dfc81e3d1726aaf1cf190b5728b4e6557abe54fcd6R184) which assign label `1` to 1 position for each of 4 (or 16) blocks and then checks if the full image has only unique labels. In general, we the labels are saved as `u32` and the lowest `k` bits encode the block id, the rest is the original label. So if number of labels/blocks is reasonable, there shouldn't be any overflow, but we can always preeemptively bump the dtype to `u64`.

> is it possible to tell dask to use several processes when applying functions to chunks?

Yes, should be double with this: https://docs.dask.org/en/latest/scheduling.html#configuration
One problem I've found is that when using processing + large images, PIL's `MAX_IMAGE_PIXELS` is reset, resulting in an error.

> did you test the lazy loading for big pngs/jpegs? I don’t have a large image like this atm to test this.

I've tried on 30k x 30k jpeg and found a bug, thanks! (seems like `pims` loads the array when fetching metadata, I will either delay it as well or find another way).

hspitzer commented:
> > I've tried on 30k x 30k jpeg and found a bug, thanks! (seems like pims loads the array when fetching metadata, I will either delay it as well or find another way).
> 
> Actually, I wanted to talk to you about this @hspitzer - currently, the ugly code in `_io.py` stems from handling tiffs. In our case,
> the `visium_hne_image_crop.tiff` is saved as `(3, 7272, 7272, 1)`, so 3 frames, 1 channel whereas the `visium_hne_image_crop.tiff` is saved as `(1, 3527, 3527, 3)`, 1 frame, 3 channels. What about reading tiffs always as `Z, Y, X, C` (+ add an option to swap `C <-> Z` in the `ImageContainer`? This would be backwards incompatible, since we save the fluo image as it has 3 channels though.

Hm, this is difficult. I just checked an original fluorescence tiff and and original h&e tiff. The fluo tiff has 3 pages and 1 channel. The h&e tiff has 1 page and 3 channels. This makes sense if you consider what is saved in the image, the fluo image encodes a different kind of information in the pages (multiple markers) than the h&e image encodes in the channels (RGB). We just decided to handle both of these types of information in the same way by mapping it to the channels dimension -- and I think this is a good option and we should make sure that if you have these types of images they are read "correctly" by default. 

I'm not sure how 3D or z-stacks data would look like, as I haven't seen examples of this saved in a single tiff -- if its saved in several files then its no problem to handle this for us. I think the code you have right now tries to disambiguate very well if there are several z stored in the same tiff. It might be a bit ugly, but as I said above, I would not change the default behaviour for fluo tiffs with one z. Imo the different pages should still be read in different channels. 

I'm very happy to discuss though, if you have a different opinion!

hspitzer commented:
> > and I think this is a good option and we should make sure that if you have these types of images they are read "correctly" by default.
> 
> Ok, then I'd propose to add an option when reading images called e.g. `determine_dims: Literal['auto', 'default'] or : Literal['prefer_channels', 'prefer_z', 'default']` where `default` is to keep the dims as they are in the tiff (i.e. `3, y, x, 1` for fluo, `1, y, x, 3` for H&E), `auto/prefer_channels` is what's done now and `prefer_z` would be like `prefer_channels`, but in the Z-dim.
> If you agree, I'd go for the 2nd option (`prefer_channels, prefer_z, default`), since it's more flexible and not too much work.

This is a great proposal, I love it! This makes it very clear what is happening, and the user can also decide on their own what they would like to use. Lets go for option 2 then.

michalk8 commented:
I've also removed `as_mask: bool` from `ImageContainer.show` and replaced it with `segmentation_layer: Optional[str]` and `segmentation_alpha: float = 0.75`, it looks e.g. as follows (alpha=1.0):
![segmentation](https://user-images.githubusercontent.com/46717574/117853056-7e9cf500-b277-11eb-909a-682a3b570c23.png)

michalk8 commented:
#337 will also be handled as (still need to write tests):
```python3
adata = ...
img = sq.im.ImageContainer(...)
adata = img.crop(...).subset(adata)
```

giovp commented:
I swear I'll test it out tomorrow or friday, sorry for super late reply.

qwuick q: these artifacts are due to `map_overlap` ?
![image](https://user-images.githubusercontent.com/25887487/117981809-b10d2780-b335-11eb-8621-6b43da63c28d.png)

or, are they artefacts at all?

michalk8 commented:
> or, are they artefacts at all?

They are not artifacts of `map_overlap`, but of watershed. Artifacts of using `map_blocks`, our previous custom chunking is seen in https://github.com/theislab/squidpy/pull/324#issuecomment-829467853

hspitzer commented on squidpy/im/_container.py:
```diff
@@ -71,13 +74,13 @@
 
-    Allows for lazy and chunked reading via :mod:`rasterio` and :mod:`dask`, if the input is a *TIFF* image.
-    This class is given to all image processing functions, along with :class:`anndata.AnnData` instance, if necessary.
+    This class also llows for lazy loading and processing using :mod:`dask`, and is given to all image
```
```suggestion
    This class also allows for lazy loading and processing using :mod:`dask`, and is given to all image
```

michalk8 replied:
Done, thanks.

hspitzer reviewed (approved):
Looks really amazing! The subset function is super cool!
Good to merge for me.

I've played around with the `infer_dimensions` fn, and it seems to do the right thing in most cases, but I'd like to ask you to document its behaviour properly somewhere. I find the code a bit hard to follow with all this switching of indices, so It would be good do either add a lot of comments in it, or to have a doc somewhere that explains how dimensions are guessed. 
If its not there already, we should also add a line or two in the docstring of `add_img` to explain that dimensions are guessed (but maybe lets do that after #329, where I add explicit dims as well).

michalk8 commented:
> I find the code a bit hard to follow with all this switching of indices, so It would be good do either add a lot of comments in it, or to have a doc somewhere that explains how dimensions are guessed.

I've added a docstring for the internal method, hopefully it's clearer now. Will update the main docstring in #329 to accomodate for explicit passing of dims.

>  Good to merge for me.

Thanks for the review, merging this.
